File: libnd4j/include/ops/declarable/helpers/cpu/toggle_bits.cpp
/* ******************************************************************************
 *
 *
 * This program and the accompanying materials are made available under the
 * terms of the Apache License, Version 2.0 which is available at
 * https://www.apache.org/licenses/LICENSE-2.0.
 *
 *  See the NOTICE file distributed with this work for additional
 *  information regarding copyright ownership.
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
 * License for the specific language governing permissions and limitations
 * under the License.
 *
 * SPDX-License-Identifier: Apache-2.0
 ******************************************************************************/

//
// @author raver119@gmail.com
//
#include <helpers/BitwiseUtils.h>
#include <ops/declarable/helpers/toggle_bits.h>
#if NOT_EXCLUDED(OP_toggle_bits)
namespace sd {
namespace ops {
namespace helpers {
template <typename T>
static void toggle_bits__(NDArray* in, NDArray* out) {
  auto lambda = LAMBDA_T(_x) { return BitwiseUtils::flip_bits(_x); };

  in->applyLambda<T>(lambda, out);
}

void __toggle_bits(sd::LaunchContext* context, NDArray* in, NDArray* out) {
  BUILD_SINGLE_SELECTOR(in->dataType(), toggle_bits__, (in, out), SD_INTEGER_TYPES);
}
}  // namespace helpers
}  // namespace ops
}  // namespace sd
#endif

File: nd4j/nd4j-backends/nd4j-api-parent/nd4j-api/src/main/java/org/nd4j/linalg/api/ops/custom/ToggleBits.java
/* ******************************************************************************
 * Copyright (c) 2019 Konduit K.K.
 *
 * This program and the accompanying materials are made available under the
 * terms of the Apache License, Version 2.0 which is available at
 * https://www.apache.org/licenses/LICENSE-2.0.
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
 * License for the specific language governing permissions and limitations
 * under the License.
 *
 * SPDX-License-Identifier: Apache-2.0
 ******************************************************************************/
package org.nd4j.linalg.api.ops.custom;

import lombok.NonNull;
import org.nd4j.autodiff.samediff.SDVariable;
import org.nd4j.autodiff.samediff.SameDiff;
import org.nd4j.base.Preconditions;
import org.nd4j.linalg.api.buffer.DataType;
import org.nd4j.linalg.api.ndarray.INDArray;
import org.nd4j.linalg.api.ops.DynamicCustomOp;

import java.util.Collections;
import java.util.List;

public class ToggleBits extends DynamicCustomOp {

    public ToggleBits() {}

    public ToggleBits(@NonNull INDArray input, INDArray output) {
        this(input);
        if (output != null) {
            addOutputArgument(output);
        }
    }

    public ToggleBits(@NonNull INDArray input) {
        addInputArgument(input);
    }

    public ToggleBits(@NonNull SameDiff sameDiff, @NonNull SDVariable input) {
        super("", sameDiff, new SDVariable[]{input});
    }

    @Override
    public String opName() {
        return "toggle_bits";
    }

    @Override
    public String tensorflowName() {
        return "Invert";
    }

    @Override
    public List<DataType> calculateOutputDataTypes(List<DataType> inputDataTypes){
        int n = args().length;
        Preconditions.checkState(inputDataTypes != null && inputDataTypes.size() == n, "Expected %s input data types for %s, got %s", n, getClass(), inputDataTypes);
        return Collections.singletonList(inputDataTypes.get(0));
    }
}
